Lib/asyncore.py

```diff
@@ -143,22 +143,17 @@ def poll(timeout=0.0, map=None):
 
         r, w, e = select.select(r, w, e, timeout)
 
-        for fd in r:
-            obj = map.get(fd)
-            if obj is None:
-                continue
-            read(obj)
+        # bpo-30931: Get objects from map before calling handlers,
+        # since handlers can modify map
+        robjs = [map[fd] for fd in r]
+        wobjs = [map[fd] for fd in w]
+        eobjs = [map[fd] for fd in e]
 
-        for fd in w:
-            obj = map.get(fd)
-            if obj is None:
-                continue
+        for obj in robjs:
+            read(obj)
```

Comment:
And now we try to read from close dispatcher.

I think we should wait for https://github.com/python/cpython/pull/2804, and do:

    for obj in ...:
        if obj.closing:
            continue
        read(obj)

And same for write/_exception.